### ISSUE TITLE
[BUG] cia checksums are different

### ISSUE THREAD

**Issue by Liger0** [d0k3/GodMode9#595]
**So you want to report a bug?**

I dumped some games and Updates/DLC as standard cia.
I compared the m5d and sha-1 to the ones of a site that provides correct ones and also to the ones of a my friend's dump and they all are different from each other.
The same games, dumped as .3ds have the same m5d/sha1.
The dlc are the same version, for example for alpha shapphire I tested v1.4 v7280.
In my case the sha1 was 9F45...CC3 (the . are for the mid values), while the site reports as correct sha-1 value 
A43794688A7B7D76A1A2B3D981C32364D2A9D96B

Since they are standard cia I don't see why they differ.

**Info about your system**
Include this info to make our work easier:
* Console type (O3DS/N3DS): new 3ds XL
* Bootloader (boot9strap/fastboot3ds): B9S

**Have you actually read this?**
[Yes] I have read the information above

**Comment by urherenow** [comment]
Because a .cia is the format when you get something from the eshop, which has a ticket meant for YOUR account on YOUR 3ds. 

There are very few exceptions to this, and those are referred to as “legit” .cia. Those are games that came bundled with a console (can’t have a ticket in there married to YOUR account, if Nintendo doesn’t know who’s going to buy this particular console).

Tl:dr

Not a bug. Please close.

**Comment by d0k3** [comment]
@urherenow - @Liger0 refers to CIAs created via cart dumps here, so you may be on to something.

@Liger0 - can you name that site? A site just containing checksums is not illegal. Is a different story if there is also piracy there.

Also keep in mind, there are multiple ways to create a CIA from a cart dump. GM9 has the most correct one (or so the author says), and that site may be using some old tool.

Expect a test build tomorrow, I think we already have a fix for this on staging.

**Comment by urherenow** [comment]
Well, DLC and updates were mentioned. Those are 100% tied to an account (or zeroed out somehow if not legit).

**Comment by d0k3** [comment]
@Liger0 - can you please use this test build to test everything (that means also the differing CIA builds with your friend):

https://f.secretalgorithm.com/oWubO/godmode9.firm

**Comment by Liger0** [comment]
[The site](https://datomatic.no-intro.org/) only contains .3ds checksums unfortunately (which corresponds to all my dumped games), it only provides .cia checksums for the updates/DLC, but if you say update cias are all different then it's not a useful info.
I dumped my european pokemon Y and pokemon Sun cartridges with the godmode9.firm you just provided and the values were those

0004000000055E00_v00 pokemon Y.cia 
SHA-1: 96D8168569E876C7E7F54E8E59767AD9696821F7

0004000000164800_v00 Pokemon ultra sun.cia SHA-1: 470C491C6504DD6CE5F5419F95FB094A08571646

I gave a try to some online 0004000000055E00_v00 pokemon Y.cia and 0004000000164800_v00 Pokemon ultra sun.cia (so they should be the exact same having the same code), but the sha-1 were different, altough this isn't enough info since it is dumped with old godmode9 versions.
So if anyone with a retail EU pokemon Y or Sun cardridge could try to dump them as cia with the just posted godmode9.firm he could compare his sha-1 to mine and tell if they are the same.

**Comment by urherenow** [comment]
Cartridges have unique serial numbers. You will NOT get the same hash. Unless the person who uploaded the image is an idiot, the specific cartridge ID should have been zeroed out. If you zero out your cartridge ID, perhaps you will have a matching sha1.

**Comment by Liger0** [comment]
In which offset is the cart ID stored? In this case to compare, zeroing it on both the cias should make them equal

**Comment by urherenow** [comment]
Sorry, I don't recall. Google "GATEWAY ROM PATCHER V1.2". That can zero out the Cart ID. If your downloaded one has a bunch of FFFFFF in the Cart ID (it should), then that's why you have a different hash. Honestly though, the best way to test is to dump YOUR cart, on 2 different systems. That should remove all doubt.

**Comment by d0k3** [comment]
@urherenow - the private header (from hard dumps) is not taken over into CIA files. In fact, that's not even possible. You're right in that there exist some cart dumps which still have the private header in it. GM9 made cart dumps always have it removed.

@Liger0 - take a look at this issue:
https://github.com/d0k3/GodMode9/issues/587
It's solved, and it's the exact reason why your friends and your own CIA checksums don't match. The fix will be in the next version of GM9. I can't say anything about CIAs made from other tools (some have rather obvious issues, which still keep them playable).

Take note that CIA files can't be perfect, it's complicated. CIA files made from cart dumps do contain tickets, which are custom built for everyone and require CFW (that's the standard in GM9 built CIAs). There's no recognized, official way on how these tickets are made.

tl;dr: CIAs are not an exact science. It's not the same as with classic systems cart dumps (which are easily verified via checksums).

Are there any still open questions? Anything you'd suggest to change?

**Comment by Liger0** [comment]
I got it. I'll wait for the next version and report back the results. Thank you.

**Comment by Ammako** [comment]
> Since they are standard cia I don't see why they differ.

Note that dat-o-matic checksums are from encrypted CIAs (presumably from the CDN.) GM9 (standard) dumps are decrypted by default. So from that alone you would never get the same results.

Pretty silly of them to bother keeping track of CIA checksums for anything other than legit CIAs though, I had no idea they did that. Especially since they already keep track of individual CDN contents and tmds. Must be recent.

GM9 standard dumps should be fine for everyday use, so you don't have to worry about them being bad, but if you care about making perfect dumps for archival purposes then you should make a (legit) dump (and zero out console ID and such in the ticket afterwards, so it can be installed), or dump your ticket and download using ctrcdnfetch. Only if your updates and DLC were legitimately obtained, that is.

**Comment by d0k3** [comment]
> GM9 standard dumps should be fine for everyday use, so you don't have to worry about them being bad, but if you care about making perfect dumps for archival purposes then you should make a (legit) dump (and zero out console ID and such in the ticket afterwards, so it can be installed).

@Ammako - If you zero out the console ID in the the ticket, the tickets signature becomes invalid, meaning the CIA is no more installable (and the content no more runable) without CFW. That means: yup, there is no proper way of building CIAs for general archival purpose. CIAs are either personalized to some console ID, or they have an invalid ticket signature and need CFW. The exception to this is a small handful of known legit CIAs (games that came preinstalled on some special edition 3DS consoles).

My suggestion for archival is: Do a standard CIA. It will only work with CFW, but starting from the next GM9 version, standard built CIAs will be identical for the same content, even when built from different consoles. FYI: The ticket signature in a standard built CIA is invalid, but at least it's a standard signature in a custom built ticket.

@Liger0 - I'm closing this as solved. Bug reports are always welcome, so thank you for this! If anything comes up, feel free to open a new issue.

**Comment by Ammako** [comment]
> If you zero out the console ID in the the ticket, the tickets signature becomes invalid, meaning the CIA is no more installable (and the content no more runable) without CFW.

Of course. What I'm referring to, is that a standard decrypted dump breaks tmd signature, and re-encrypting the CIA doesn't fix this.

Archival in this context isn't just for preserving playable and functional content, it's to preserve things as closely as possible to the original. So I guess maybe I should have used the term "preservation" instead. The only way to do this in GM9 is to make a legit dump, and edit the ticket for it to be installable. Then the original encryption is preserved, and tmd signature remains intact. Same as if you built a CIA from CDN, unless I'm missing something

The ticket would be different from those of course, but that in itself wouldn't make it invalid. If Nintendo wants to check for fake ticket they'll be able to do this even if you used a standard ticket template.

There isn't really any value in preserving tickets, aside from univeral legit ones. So as long as tmd and contents are intact, that's all that's really needed for archival. One way to handle this in GM9, if you were interested in handling these types of cases, would be to have a separate option for legit cia building which checks if a legit ticket is present, and then it builds a legit CIA as normal, but instead of using the legit ticket, it would use the standard GM9 fake ticket template, with title key taken from the legit ticket. Otherwise this is already doable by manually replacing the ticket yourself, or dumping the legit ticket and using ctrcdnfetch (although this won't be possible anymore in the future, once eshop servers might get taken down, so it could be good to have a way to do it offline.)

**Comment by d0k3** [comment]
I actually like that idea, and I'd be okay with adding a third way of building CIAs. Reopening this so I don't forget. Still need to think about the naming of the three options, so I don't confuse people.

**Comment by urherenow** [comment]
Legit->Bundled game
Personal->Purchased game backup
Public-> ticket template use
(Or Anon, for anonymous)


Isn’t the titlekey already used when building a cia? I’m pretty sure it’s a per-title thing, and not per-cartridge/account. Not sure why anything else would be used for a “Public” dump. I guess I don’t understand the question. I led with my suggestion for the names though. It makes sense to me, and I don’t see a reason for a 4th type.

**Comment by Ammako** [comment]
I was thinking, perhaps standard could be split into two. One which decrypts the contents if needed (what standard already does), and one which leaves any encryption alone. Both would use the same template fake ticket.

As long as we can guarantee that we're reading the correct ticket (and not, say, a deleted duplicate ticket which could contain an incorrect title key), we can be reasonably certain that the title key in the ticket will work. So it's not totally necessary to have a legit eShop ticket in order to do this.

Alternatively, I guess it could use the titlekey from encTitleKeys.bin if available, but there's always the possibility that someone has a bad file with invalid title keys in it, so I'm not sure if this would be reliable enough.

If you have stuff you've downloaded from eShop, this would maintain the original encryption with valid tmd signature. It would also work if you've got untouched CDN CIAs that you've downloaded and installed, even if you don't have legit tickets for them in ticket.db

I'm not 100% sure how title encryption works though, when you install a CIA. Never really looked into it, and I don't really have time for that right now. But hopefully this option would work even if the installed contents were 'dirty' (custom CIAs, gamecard converts, anything that doesn't have valid tmd signature.) That way, even if a less savvy user picked that option, they would still get a functional CIA out of it and it wouldn't cause any issues. I could be wrong though.

Speaking of encryption. I know that titles installed to nand need to be encrypted, or they won't work. This isn't immediately obvious, so the ability to directly build a CIA as encrypted without needing extra steps would be useful there. This is kind of a niche case, since most nand titles will usually be made into legit CIAs anyway, but it would be useful for things like dev tools or modified system titles.

**Comment by d0k3** [comment]
@Ammako - Tickets in legit CIAs have their signature validated. We can guarantee these tickets are fine. We're also already using the encTitlekeys.bin (or decTitlekeys.bin) file if it is available, as fallback when there is no ticket found or ticket is not available.

As for system titles, yeah, the standard CIA option makes little sense in most cases. We'll still leave it in. Most people creating CIAs from system software know their stuff (at least that's what I assume).

As for leaving the TMD and ecnryption intact: I added the possibility to do that. Just select the legit CIA builder. If no legit ticket is found or if the legit ticket is personalized, GM9 will now offer the generic ticket as workaround, leaving everything else as is.

With that, I'm closing this issue now.

**Comment by Ammako** [comment]
@d0k3 so, I got around to trying this. I'm a bit confused though, because CIAs built in this manner have the title key totally FF'd out in their ticket, and yet they install and run just fine, despite missing the correct title key. To me, that would imply that the CIA lacks title key crypto, but wasn't it supposed to leave all encryption intact? Despite this, GM9 recognizes TMD as legit. How does that work? Am I missing something?

**Comment by d0k3** [comment]
Sorry, didn't notice your last comment earlier. Well, what you recognized as title key is most likely something else, like the signature or ECC PublicKey. Check here:
https://www.3dbrew.org/wiki/Ticket

**Comment by Ammako** [comment]
No, it's most definitely the title key. 0x1BF-0x1CE in ticket.bin

**Comment by d0k3** [comment]
Ooof. Can you do an issue for the enc/decTitlekeys thing?

As for that CIA - does it install in FBI? You could also upload it if you want. For GM9 created data this should be okay.

**Comment by Ammako** [comment]
They both install and run perfectly fine, which is what doesn't make sense to me and tells me that either the title key crypto is not being retained, or the crypto is being re-done using FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF as the encrypted title key. Whatever it's doing though, it's not affecting tmd signatures.

<details>
    <summary>Borked:</summary>

![image](https://user-images.githubusercontent.com/43770697/132091658-9e3bfbdf-c50f-467f-9213-3fca539fec9e.png)
![image](https://user-images.githubusercontent.com/43770697/132091672-902a1c9f-b6c5-49a3-91a5-bc5f6850e202.png)

</details>

<details>
    <summary>For comparison:</summary>

![image](https://user-images.githubusercontent.com/43770697/132091685-52a6a8b8-b672-47e4-bc06-ad071d525438.png)
![image](https://user-images.githubusercontent.com/43770697/132091770-8531eb4e-f3d9-4630-b2bb-bb975c8583ed.png)

</details>

As for uploading: are you sure it's a good idea? These are perfectly functional CIAs, for copyrighted content.

**Comment by d0k3** [comment]
By uploading I meant the ticket. Sorry, I see how what I wrote was misleading now.

I do see this is actually the title need, which is very weird. Is the CIA actually encrypted? You can easily check by searching for "NCCH" in the CIA. If that's not a DSiWare title and you can't find it, then it is encrypted.

**Comment by Ammako** [comment]
It's encrypted. I'm able to decrypt it, breaking tmd signature in the process, and you can very clearly tell the difference between encrypted and decrypted data anyway (and of course NCCH and product code that become clearly exposed once decrypted.)

If you need a cia to check things on your own, as long as you have some kind of genuine ticket for a title (updates are easy and free), you can get a clean cdn cia with ctrcdnfetch. Wipe the original ticket from your ticket.db or install the cia to a different console entirely, to make sure gm9 can't just erroneously pick up the original legit ticket when re-dumping, and you should get a cia like this to play around with.

**Comment by d0k3** [comment]
Reopening this, this will need attention. Even though I still think there's something we don't see here. Also, can't check this earlier than 3 weeks from now. Don't have my dev hardware with me. So, and additional feedback from other sides is welcome, too.

**Comment by Ammako** [comment]
I'm pretty sure it's redoing titlekey crypto using FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF as the title key, rather than omiting it entirely, because if I edit the title key in the cia then the verification fails. So your code might be failing to retrieve the title key from the ticket and defaulting to all 0xFF as per     `memset(_ticket->titlekey, 0xFF, 16);`

**Comment by TimmSkiller** [comment]
Was able to reproduce the same way @Ammako said to:

1. Have a title with a legit ticket installed.
2. Dump it as a piratelegit CIA. 
3. Install it.
4. Dump it as piratelegit again.
5. Title key is filled with 0xFF.
6. Verify this CIA: 
```
TMD/Ticket: illegit/legit
Verification: passed
```
Even though it is 100% the incorrect title key.

I went a bit further, doing a bit more testing:

Hashed:
- Fully legit CIA
- Piratelegit CIA
- Piratelegit CIA with incorrect title key

Even though the last two should have the same hash, they did not.
The contents inside all three CIAs are exactly the same after being verified by hashing them.
What is NOT the same however is the titlekey in the ticket.

Both the fully legit CIA and the piratelegit CIA (not the FF titlekey one) share the same title key. The incorrect titlekey CIA however really does seem to use 0xFF. 

The reason it doesn't affect functionality is because it uses FF as the titlekey, the contents are properly encrypted using that FF key. This is why it still is able to be installed, read, en- and decrypted, played and whatnot.

This also doesn't affect the TMD validity either as the TMD only cares about a content's hash when it's **decrypted**, __**not**__ while it's encrypted with CDN encryption. It uses FF to decrypt it, hence why the hash matches since the content is still valid after all.

Somewhere along the line something is setting the title key to FF where it shouldn't. That is the real issue here.

**Comment by TimmSkiller** [comment]
Alright. After a bit of troubleshooting, I found out what the issue is.

If an illegit ticket is installed, **it will not use it**. This means that incase of already installed illegit tickets, it will continue to use the default 0xFF key that is set before any tickets/titlekeys are loaded.

Knowing this, I added some more checks to detect this and ask the user whether they want to:

- Use the installed (illegit, potentially with the correct title key) ticket
- Use the 0xFF key (illegit)

It will not allow you to build legit CIAs if you don't have any form of ticket/titlekey installed.

The way to test it is just like before; However, with the exception of the new check that was added.
In my own testing, this seems to work perfectly, with the expected output every time.
I committed these changes to my [fork](https://github.com/TimmSkiller/GodMode9).

Here are some test builds:
[GodMode9.firm.zip](https://github.com/d0k3/GodMode9/files/7114345/GodMode9.firm.zip)
[GodMode9_dev.firm.zip](https://github.com/d0k3/GodMode9/files/7114346/GodMode9_dev.firm.zip)

Let me know what you think.

**Comment by Ammako** [comment]
imo the choice is not useful, because as long as the title is functional, it doesn't really matter if the title key in the ticket could be wrong, because it'll build a functional cia regardless.

That being said, if you've got intact tmd, there's little reason to believe the title key in the ticket would be incorrect, either way. Aside from this case here where 0xFF times 16 ended up being used ;p

**Comment by TimmSkiller** [comment]
You could also just make it use whatever is installed. It won't really make a difference.

**Comment by Ammako** [comment]
All I can see is that the [original implementation](https://github.com/d0k3/GodMode9/commit/a7ce125d7fde599513ba65c45e7b393b4516f4cb) meant to pull the title key from the installed ticket and graft it onto the generic ticket.

At some point some [major refactoring](https://github.com/d0k3/GodMode9/commit/f9408a9c10a8d81765d128bb3fc6e5c187b3e080) was done to this part of the code, which may or may not have broken the originally-intended functionality, but I haven't gotten around yet to checking if the original implementation worked, vs the new one.

I'm always wary of code refactorings, you see 😉 good way to introduce regressions if everything isn't thoroughly tested afterwards.

**Comment by mariomadproductions** [comment]
I'm not 100% sure if the CIAs in the "Nintendo - Nintendo 3DS (Digital) (Updates and DLC)" dat are made in a consistent manner. Its basically a third party dat and I'm not sure the creator/gatherer of the CIAs in it is around anymore. "(Digital) (CDN)" is the main (but WIP) dat, which just has the raw CDN contents.

**Comment by d0k3** [comment]
@Liger0 @Ammako @TimmSkiller - here's a test build that fixes (hopefully) the issue with less code changes and also, similar bugs in other places.
[GodMode9-v2.0.0-28-g27e31657-20210926211820.zip](https://github.com/d0k3/GodMode9/files/7232092/GodMode9-v2.0.0-28-g27e31657-20210926211820.zip)

Could you please test?

Thanks @Ammako, too, you were right about the FindTitleKey function no more checking the internal ticket database.

**Comment by Ammako** [comment]
At first glance it looks good, title key is there and the verification passes.
